File: backend/internal/utils/portfolio.go
package utils

import (
	"fintu-tracking-backend/internal/models"
	"sort"

	"github.com/shopspring/decimal"
)

// CalculateHoldings calculates current holdings from trade history
func CalculateHoldings(trades []models.Trade) map[string]models.Holding {
	holdings := make(map[string]models.Holding)

	// Group trades by ticker
	tradesByTicker := make(map[string][]models.Trade)
	for _, trade := range trades {
		tradesByTicker[trade.Ticker] = append(tradesByTicker[trade.Ticker], trade)
	}

	// Calculate holdings for each ticker
	for ticker, tickerTrades := range tradesByTicker {
		// Sort by date ascending
		sort.Slice(tickerTrades, func(i, j int) bool {
			return tickerTrades[i].Date.Before(tickerTrades[j].Date)
		})

		totalQuantity := decimal.Zero
		totalCost := decimal.Zero

		for _, trade := range tickerTrades {
			quantity, _ := decimal.NewFromString(trade.Quantity)
			total, _ := decimal.NewFromString(trade.Total)

			if trade.Side == "buy" {
				totalQuantity = totalQuantity.Add(quantity)
				totalCost = totalCost.Add(total)
			} else {
				// Sell - reduce quantity and cost proportionally
				if totalQuantity.GreaterThan(decimal.Zero) {
					avgCost := totalCost.Div(totalQuantity)
					totalQuantity = totalQuantity.Sub(quantity)
					totalCost = totalQuantity.Mul(avgCost)
				}
			}
		}

		// Only include if we still hold shares
		if totalQuantity.GreaterThan(decimal.Zero) {
			avgCost := totalCost.Div(totalQuantity)
			holdings[ticker] = models.Holding{
				Ticker:              ticker,
				Quantity:            totalQuantity.String(),
				AvgCost:             avgCost.String(),
				TotalInvested:       totalCost.String(),
				MarketValue:         "0",
				UnrealizedPL:        "0",
				UnrealizedPLPercent: "0",
			}
		}
	}

	return holdings
}

// UpdateHoldingsWithPrices updates holdings with current market prices
func UpdateHoldingsWithPrices(holdings map[string]models.Holding, prices map[string]string) map[string]models.Holding {
	updatedHoldings := make(map[string]models.Holding)

	for ticker, holding := range holdings {
		if priceStr, ok := prices[ticker]; ok {
			quantity, _ := decimal.NewFromString(holding.Quantity)
			price, _ := decimal.NewFromString(priceStr)
			totalInvested, _ := decimal.NewFromString(holding.TotalInvested)

			marketValue := quantity.Mul(price)
			unrealizedPL := marketValue.Sub(totalInvested)
			unrealizedPLPercent := unrealizedPL.Div(totalInvested).Mul(decimal.NewFromInt(100))

			updatedHoldings[ticker] = models.Holding{
				Ticker:              ticker,
				Quantity:            holding.Quantity,
				AvgCost:             holding.AvgCost,
				TotalInvested:       holding.TotalInvested,
				MarketValue:         marketValue.String(),
				UnrealizedPL:        unrealizedPL.String(),
				UnrealizedPLPercent: unrealizedPLPercent.String(),
			}
		} else {
			updatedHoldings[ticker] = holding
		}
	}

	return updatedHoldings
}

File: lib/portfolio.ts
import { Decimal } from "./decimal"
import type { Trade, Holding } from "./types"

export function calculateHoldings(trades: Trade[]): Map<string, Holding> {
  const holdings = new Map<string, Holding>()

  // Group trades by ticker
  const tradesByTicker = trades.reduce(
    (acc, trade) => {
      if (!acc[trade.ticker]) {
        acc[trade.ticker] = []
      }
      acc[trade.ticker].push(trade)
      return acc
    },
    {} as Record<string, Trade[]>,
  )

  // Calculate holdings for each ticker
  for (const [ticker, tickerTrades] of Object.entries(tradesByTicker)) {
    // Sort by date ascending
    const sortedTrades = tickerTrades.sort((a, b) => new Date(a.date).getTime() - new Date(b.date).getTime())

    let totalQuantity = new Decimal(0)
    let totalCost = new Decimal(0)

    for (const trade of sortedTrades) {
      const quantity = new Decimal(trade.quantity)
      const total = new Decimal(trade.total)

      if (trade.side === "buy") {
        totalQuantity = totalQuantity.add(quantity)
        totalCost = totalCost.add(total)
      } else {
        // Sell - reduce quantity and cost proportionally
        if (totalQuantity.gt(0)) {
          const avgCost = totalCost.div(totalQuantity)
          totalQuantity = totalQuantity.sub(quantity)
          totalCost = totalQuantity.mul(avgCost)
        }
      }
    }

    // Only include if we still hold shares
    if (totalQuantity.gt(0)) {
      const avgCost = totalCost.div(totalQuantity)
      holdings.set(ticker, {
        ticker,
        quantity: totalQuantity.toString(),
        avgCost: avgCost.toString(),
        totalInvested: totalCost.toString(),
        marketValue: "0",
        unrealizedPL: "0",
        unrealizedPLPercent: "0",
      })
    }
  }

  return holdings
}

export function updateHoldingsWithPrices(
  holdings: Map<string, Holding>,
  prices: Map<string, string>,
): Map<string, Holding> {
  const updatedHoldings = new Map<string, Holding>()

  for (const [ticker, holding] of holdings) {
    const price = prices.get(ticker)
    if (price) {
      const quantity = new Decimal(holding.quantity)
      const avgCost = new Decimal(holding.avgCost)
      const marketValue = quantity.mul(price)
      const totalInvested = new Decimal(holding.totalInvested)
      const unrealizedPL = marketValue.sub(totalInvested)
      const unrealizedPLPercent = unrealizedPL.div(totalInvested).mul(100)

      updatedHoldings.set(ticker, {
        ...holding,
        marketValue: marketValue.toString(),
        unrealizedPL: unrealizedPL.toString(),
        unrealizedPLPercent: unrealizedPLPercent.toString(),
      })
    } else {
      updatedHoldings.set(ticker, holding)
    }
  }

  return updatedHoldings
}
